### DISCUSSION TITLE
Cannot dynamically add fields alongside removal.

### DISCUSSION THREAD

**Discussion by mjsir911** [encode/django-rest-framework#8528] [Potential Issue]
Hi,

per the [dynamically modifying fields](https://www.django-rest-framework.org/api-guide/serializers/#dynamically-modifying-fields), I would like to modify the `source` attribute of a field at runtime via `__init__`

```python
class TestSerializer(ModelSerializer):
    generic_test = IntegerField()
    def __init__(self, *args, direction, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields['something'].source = f"{direction}_test"
    class Meta:
         fields = [ "generic_test" ]
```

Obviously I can't modify a class attribute for a single instance I want modified, so I move the attribute outside of class initialization:

```python
class TestSerializer(ModelSerializer):
    def __init__(self, *args, direction, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields['generic_test'] = IntegerField(source=f"{direction}_test")
    class Meta:
         fields = [ "generic_test" ]
```

but this errors out due to the field not being a class attribute at class-definition-time.

The current workaround:

```python
class TestSerializer(ModelSerializer):
    generic_test = IntegerField()
    def __init__(self, *args, direction, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields['generic_test'] = IntegerField(source=f"{direction}_test")
    class Meta:
         fields = [ "generic_test" ]
```

this repeats unnecessary code and doesn't signal properly that the initial class attribute is going to always be overwritten.

Is the documentation's "[Dynamically modifying fields](https://www.django-rest-framework.org/api-guide/serializers/#dynamically-modifying-fields)" *only* for filtering out subsets of fields? There's no real way to modify or add fields in `__init__` currently.

**Comment by lovelydinosaur** [comment]
I don't think I would trust that setting `.source = ...` would necessarily work as expected.
Perhaps it does, perhaps it doesn't.

I haven't looked into it, but it's certainly not intentional usage.

Personally if I really needed to do this I'd probably just change my outlook to be okay with this as a constraint.

```python
class TestSerializer(ModelSerializer):
    generic_test = IntegerField()  # Note that we dynamically update this field in `__init__`

    def __init__(self, *args, direction, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields['generic_test'] = IntegerField(source=f"{direction}_test")

    class Meta:
         fields = [ "generic_test" ]
```

this "repeats unnecessary code" but does "signal properly that the initial class attribute is going to always be overwritten."

Not perfect. But perfectly okay. 😌

**Reply by mjsir911** [reply]
Yeah, this is basically what I ended up doing with a `DummyField()` for better signalling, but I would consider it a bug that I can't assign new `self.field[]`s without causing errors.

The documentation makes it sound like more can be done with `self.fields` than *just* filtering, even if that's the only example provided.

**Reply by mjsir911** [reply]
Even something like an emergency hatch `_declared_fields = [ 'generic_test' ]` on the class would be helpful. This is a super fringe edge case and just providing a sane *intended* way to do this would be appreciated.
